src/main/kotlin/io/kuzzle/sdk/coreClasses/json/ConcurrentHashMapTypeAdapter.kt

```diff
@@ -0,0 +1,127 @@
+package io.kuzzle.sdk.coreClasses.json
+
+import com.google.gson.JsonSyntaxException
+import com.google.gson.TypeAdapter
+import com.google.gson.internal.LazilyParsedNumber
+import com.google.gson.stream.JsonReader
+import com.google.gson.stream.JsonToken
+import com.google.gson.stream.JsonWriter
+import io.kuzzle.sdk.coreClasses.maps.KuzzleMap
+import java.io.IOException
+import java.util.*
+import java.util.concurrent.ConcurrentHashMap
+
+class ConcurrentHashMapTypeAdapter : TypeAdapter<ConcurrentHashMap<String?, Any?>?>() {
+  @Throws(IOException::class)
+  override fun write(out: JsonWriter, map: ConcurrentHashMap<String?, Any?>?) {
+    if (map == null) {
+      out.nullValue()
+    } else {
+      out.beginObject()
+      val iterator = map.entries.iterator()
+      while (iterator.hasNext()) {
+        val entry = iterator.next()
+        out.name(entry.key)
+        writeObject(out, entry.value)
+      }
+      out.endObject()
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun read(`in`: JsonReader): ConcurrentHashMap<String?, Any?>? {
+    val peek = `in`.peek()
+    if (peek == JsonToken.NULL) {
+      `in`.nextNull()
+      return null
+    } else if (peek == JsonToken.BEGIN_OBJECT) {
+      val map = KuzzleMap()
+      var key: Any
+      var value: Any?
+      `in`.beginObject()
+      while (`in`.hasNext()) {
+        key = `in`.nextName()
+        value = readObject(`in`)
+        if (!map.containsKey(key)) {
+          map.put(key, value)
+        } else {
+          throw JsonSyntaxException("duplicate key: $key")
+        }
+      }
+      `in`.endObject()
+      return map
+    }
+    return null
+  }
+
+  @Throws(IOException::class)
+  private fun writeObject(out: JsonWriter, value: Any?) {
+    if (value is Number) {
```

Comment:
Kotlin has a pattern matching system that might work better for this use case (https://kotlinlang.org/docs/reference/control-flow.html#when-expression)
```kotlin
when (value) {
  is Number -> out.value(value as Number?)
  is Boolean -> out.value(value as Boolean)
  is ArrayList<*> -> { // you can declare scoped branch
    // The parsing logic you made...
  }
  else -> doSomethingElse() // You can also use 'else' 
}
```

Reply:
I know about this pattern, I use it somewhere. I didn't write everything, I took the source from Java and let intellij transform everything, the goal being to make this SDK asap

Reply:
Oh ok nevermind 🙂